docs/coding-and-design-guidelines.md

```diff
@@ -0,0 +1,50 @@
+This document lists identified coding and design guidelines that should be followed when working on ServiceControl. In case of conflicts with other coding and design guidelines, this document should take precedence when working on ServiceControl.
+
+## Prefer Microsoft abstractions
+
+Microsoft maintains a number of abstractions for common cross-cutting concerns in the .NET common library and in the `Microsoft.Extensions.*` namespaces. Where such an abstraction exists, we prefer to use it over an abstraction provided by a different assembly (including one of our own). 
+
+This makes it easier for us to isolate the application from 3rd party dependencies by relying on fairly stable abstractions. It also helps to keep parts of the application isolated from each other (i.e. being able to run the embedded database for maintenance mode without starting the NServiceBus endpoint).
+
+For example:
+
+- **Prefer IHostBuilder extensions over NServiceBus features**: Unless the code being added is specifically altering the NServiceBus endpoint, it makes sense to add new ServiceControl code directly as extensions to `IHostBuilder` rather than as NServiceBus `Feature` implementations. Some existing ServiceControl features may still be using the `Feature` abstraction to register its components. If these features are not modifying the NServiceBus endpoint, then they should be migrated to `IHostBuilder` extensions over time.
+- **Prefer Hosted Services over NServiceBus FeatureStartupTasks**: ServiceControl includes a lot of background tasks and these are both abstractions for building background tasks. `FeatureStartupTask` implementations are tied to the lifecycle of an endpoint. In general, we prefer to use the `IHostedService` abstraction which is tied to the lifecycle of the host application. NOTE: Hosted Services are started in the order that they are registered which allows us to better control the startup sequence. They are shut down in the reverse order.
+- **Prefer IServiceCollection over IConfigureComponents or IContainerBuilder**: Where possible we'd rather use the generic Microsoft DI abstractions over the NServiceBus one (IConfigureComponents) or the Autofac one (IContainerBuilder).
+  - When registering components from within an NServiceBus Feature, it still  makes sense to use IConfigureComponents but we should consider whether it makes sense to move the code out of NServiceBus features
```

Comment:
Does it make sense to add a note here to mention that this will be “solved” when SC will be migrated to v8, since v8 uses IServiceCollection by default?